core/src/main/java/org/apache/brooklyn/enricher/stock/reducer/Reducer.java

```diff
@@ -0,0 +1,187 @@
+    @SetFromFlag("transformation")
+    public static final ConfigKey<String> REDUCER_FUNCTION_TRANSFORMATION = ConfigKeys.newStringConfigKey("enricher.reducerFunction.transformation",
+        "A string matching a pre-defined named reducer function, such as joiner");
+    public static final ConfigKey<Map<String, Object>> PARAMETERS = ConfigKeys.newConfigKey(new TypeToken<Map<String, Object>>() {}, "enricher.reducerFunction.parameters", 
+        "A map of parameters to pass into the reducer function");
+   
+    protected Entity producer;
+    protected List<AttributeSensor<S>> subscribedSensors;
+    protected Sensor<T> targetSensor;
+    protected Function<List<S>, T> reducerFunction;
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
+    @Override
+    public void setEntity(EntityLocal entity) {
+        super.setEntity(entity);
+        Preconditions.checkNotNull(getConfig(SOURCE_SENSORS), "source sensors");
+
+        this.producer = getConfig(PRODUCER) == null ? entity : getConfig(PRODUCER);
+        List<AttributeSensor<S>> sensorListTemp = Lists.newArrayList();
+
+        for (Object sensorO : getConfig(SOURCE_SENSORS)) {
+            AttributeSensor<S> sensor = Tasks.resolving(sensorO).as(AttributeSensor.class).timeout(ValueResolver.REAL_QUICK_WAIT).context(producer).get();
+            Optional<? extends Sensor<?>> foundSensor = Iterables.tryFind(sensorListTemp, 
+                    SensorPredicates.sensorNameEqualTo(sensor.getName()));
+            
+            if(!foundSensor.isPresent()) {
+                sensorListTemp.add(sensor);
+            }
+        }
+        
+        String reducerName = config().get(REDUCER_FUNCTION_TRANSFORMATION);
+        Function<List<S>, T> reducerFunction = (Function) config().get(REDUCER_FUNCTION);
+        if(reducerFunction == null){
+            Map<String, ?> parameters = config().get(PARAMETERS);
+            reducerFunction = createReducerFunction(reducerName, parameters);
+        }
+
+        this.reducerFunction = reducerFunction;
+        Preconditions.checkState(sensorListTemp.size() > 0, "Nothing to reduce");
+
+        for (Sensor<S> sensor : sensorListTemp) {
+            subscribe(producer, sensor, this);
+        }
+
+        subscribedSensors = ImmutableList.copyOf(sensorListTemp);
+    }
+
+    protected abstract Function<List<S>, T> createReducerFunction(String reducerName, Map<String, ?> parameters);
+    
+    @SuppressWarnings("unchecked")
+    @Override
+    public void onEvent(SensorEvent<Object> event) {
+        Sensor<T> destinationSensor = (Sensor<T>) getConfig(TARGET_SENSOR);
+
+        List<S> values = Lists.newArrayList();
+
+        for (AttributeSensor<S> sourceSensor : subscribedSensors) {
+            S resolvedSensorValue = entity.sensors().get(sourceSensor);
+            if (resolvedSensorValue == null) {
+                // only apply function if all values are resolved
```

Comment:
Is this always the behaviour we want - to not call the function if any of the values are null? My gut feel is to leave that up to the function (it can filter nulls if it wants).

Doing it this way could lead to strange behaviour - e.g. sensor X and Y get set to "a" and "b", then X gets set to `null`. The function will not be notified about the change of X.

I don't have any particular use-cases in mind where we'd want to know about the null - it's just a gut feel that leaving it up to the function gives more flexibility.